pwiz_tools/Shared/Common/SystemUtil/TrackAttribute.cs

```diff
@@ -0,0 +1,318 @@
+        }
+
+        public ObjectGroup<T> OldObjectGroup
+        {
+            get { return ObjectGroup<T>.Create(OldObject, OldParentObject, OldRootObject); }
+        }
+
+        public ObjectGroup<T> NewObjectGroup
+        {
+            get { return ObjectGroup<T>.Create(NewObject, NewParentObject, NewRootObject); }
+            private set { NewObject = value.Object; NewParentObject = value.ParentObject; NewRootObject = value.RootObject; }
+        }
+    }
+
+    public class ObjectPair<T> : Immutable
+    {
+        public ObjectPair(T oldObject, T newObject)
+        {
+            OldObject = oldObject;
+            NewObject = newObject;
+        }
+
+        public static ObjectPair<T> Create(T oldObj, T newObj)
+        {
+            return new ObjectPair<T>(oldObj, newObj);
+        }
+
+        public ObjectPair<T> ChangeOldObject(T oldObject)
+        {
+            return ChangeProp(ImClone(this), im => im.OldObject = oldObject);
+        }
+
+        public ObjectPair<T> ChangeNewObject(T newObject)
+        {
+            return ChangeProp(ImClone(this), im => im.NewObject = newObject);
+        }
+
+        public ObjectPair<S> Transform<S>(Func<T, S> func)
+        {
+            return Transform(func, func);
+        }
+
+        public ObjectPair<S> Transform<S>(Func<T, S> oldFunc, Func<T, S> newFunc)
+        {
+            return ObjectPair<S>.Create(oldFunc(OldObject), newFunc(NewObject));
+        }
+
+        public bool Equals()
+        {
+            return Equals(OldObject, NewObject);
+        }
+
+        public bool ReferenceEquals()
+        {
+            return ReferenceEquals(OldObject, NewObject);
+        }
+
+        public T OldObject { get; private set; }
+        public T NewObject { get; private set; }
+    }
+
+    public class ObjectPair
+    {
+        public static ObjectPair<T> Create<T>(T oldObj, T newObj)
+        {
+            return ObjectPair<T>.Create(oldObj, newObj);
+        }
+    }
+
+    public class ObjectGroup<T>
+    {
+        public ObjectGroup(T obj, T parentObject, T rootObject)
+        {
+            Object = obj;
+            ParentObject = parentObject;
+            RootObject = rootObject;
+        }
+
+        public static ObjectGroup<T> Create(T obj, T parentObject, T rootObject)
+        {
+            return new ObjectGroup<T>(obj, parentObject, rootObject);
+        }
+
+        public T Object { get; private set; }
+        public T ParentObject { get; private set; }
+        public T RootObject { get; private set; }
+    }
+
+    public interface IAuditLogObject
+    {
+        string AuditLogText { get; }
+        // Determines whether the AuditLogText is a name or a string representation
+        // of the object
+        bool IsName { get; }
+    }
+
+    public interface IAuditLogComparable
+    {
+        object GetDefaultObject(ObjectInfo<object> info);
+    }
+
+    public abstract class DefaultValues
```

Comment:
It looks like this class is being used for something similar to what "PropertyDescriptor.CanResetValue" does.
I wonder if there is any way we could have leveraged all of the PropertyDescriptor stuff.
Specifically, there's the class "System.ComponentModel.TypeDescriptor", and for things whose properties are more complicated than what reflection would reveal, you can register a TypeDescriptorProvider.

Reply:
Thanks for the suggestions, I implemented most of them (not pushed yet). I've never used type descriptors before, I'll look into it.